File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if len(matrix) == 0:
            return []
        if len(matrix) == 1:
            return matrix[0]
        
        result, inner_matrix = self.pop_border(matrix) 
        return result + self.spiralOrder(inner_matrix)
        
    def pop_border(self, matrix: List[List[int]]) -> (List, List[List[int]]):
        border = matrix.pop(0)
        
        if len(matrix[0]) == 0:
            return border, [[]]
        
        for i in range(0, len(matrix)-1):
            right_elem = matrix[i].pop(len(matrix[i])-1)
            border.append(right_elem)
        
        reverse_bottom = matrix.pop(len(matrix)-1)[::-1]
        border.extend(reverse_bottom)
        
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return border, matrix
        
        for j in range(len(matrix)-1, 0, -1):
            left_elem = matrix[j].pop(0)
            border.append(left_elem)
        
        return border, matrix  

File: 0054-spiral-matrix/0054-spiral-matrix.go
func spiralOrder(matrix [][]int) []int {
    if len(matrix) == 0 {
        return []int{}
    }
    
    if len(matrix) == 1 {
        return matrix[0]
    }
    
    border, matrix := popBorder(matrix)
    return append(border, spiralOrder(matrix)...)
}

func popBorder(matrix [][]int) ([]int, [][]int) {
    // pop the top row
    result := matrix[0]
    matrix = matrix[1:]
    if len(matrix[0]) == 0 {
        return result, matrix
    }
    
    // pop the right column top-down
    rightIdx := len(matrix[0]) - 1
    for i := 0; i < len(matrix) - 1; i++ {
        elem := matrix[i][rightIdx]
        matrix[i] = matrix[i][:rightIdx]
        result = append(result, elem)
    }
    
    // pop the bottom row and reverse
    bottom := matrix[len(matrix)-1]
    matrix = matrix[:len(matrix)-1]
    for i := len(bottom)-1; i >= 0; i-- {
        result = append(result, bottom[i])
    }
    if len(matrix) == 0 || len(matrix[0]) == 0 {
        return result, matrix
    }
    
    // pop the left column bottom-up
    for i := len(matrix) - 1; i >= 0; i-- {
        elem := matrix[i][0]
        matrix[i] = matrix[i][1:]
        result = append(result, elem)
    }
    
    return result, matrix
}